### PR TITLE
Fix `extract_slice`  causing compilation errors

### PR DESCRIPTION
**Pull request by IanWood1**
Addresses [this SHARK-TestSuite issue](https://github.com/iree-org/iree/issues/17455) related to `ExtractSliceOp` causing compilation failures. The issue originated from a pattern of `dequantize-like` -> `tensor.extract_slice` -> `quantize-like` (within a dispatch). This means that the `tensor.extract_slice` op was operating on f32 instead of the quantized i8 types. MLIR doesn't handle bufferization on extractslice/differing bit-width casts (see [mlir/*/EmptyTensorElimination.cpp](https://github.com/llvm/llvm-project/blob/6b7b05b55926f1fecae34312a89c4271def7aa3a/mlir/lib/Dialect/Bufferization/Transforms/EmptyTensorElimination.cpp#L118)), causing stack allocations in the middle of dispatches.

This is addressed by adding `linalg.generic` ops before and after the extract_slice op that cast back to the low bit-width type, eliminating the f32 tensors (after getting merged with the `dequantize-like/quantize generic`

TODOs:
- Add iree tests (although passes SHARK-TestSuite, add iree tests)
- Fix handling/matching of element types

Related issues:
https://github.com/nod-ai/SHARK-TestSuite/issues/182
https://github.com/nod-ai/SHARK-Turbine/issues/683

### CODE REVIEW

File: compiler/src/iree/compiler/Dialect/Flow/Transforms/FusionPreprocessing.cpp [hunk 1]
```diff
@@ -142,6 +142,104 @@ struct FoldSuccessiveTensorInsertSliceOps
...
+        sliceOp.getLoc(), sourceWideType.getShape(),
+        sourceWideType.getElementType());
+
+    auto newGeneric = rewriter.create<linalg::GenericOp>(
+        sliceOp.getLoc(), sourceWideType, sourceGenericOp.getResults(), emptyOp,
+        sourceGenericOp.getIndexingMapsArray(),
+        sourceGenericOp.getIteratorTypesArray(),
+        [&](OpBuilder &nestedBuilder, Location loc, ValueRange args) {
+          // Custom region for f32 -> i8 conversion
+          auto castOp = nestedBuilder.create<arith::FPToSIOp>(
```
Comment (zjgarvey) [inline compiler/src/iree/compiler/Dialect/Flow/Transforms/FusionPreprocessing.cpp]: How does modification by the quantization scale come into play here? If you just cast the fp to si, I can't imagine you will retain any useful information. 

Is it possible to copy the payload of a generic op, but modify the other information?

File: compiler/src/iree/compiler/Dialect/Flow/Transforms/FusionPreprocessing.cpp [hunk 2]
```diff
@@ -142,6 +142,104 @@ struct FoldSuccessiveTensorInsertSliceOps
...
+    }
+
+    auto sourceResult = sourceGenericOp.getResult(0);
+    auto sourceNarrowType = dyn_cast<RankedTensorType>(sourceResult.getType());
+    auto sourceWideType = cast<RankedTensorType>(
+        sourceGenericOp.getDpsInputs().front().getType());
+    auto extractedNarrowType = RankedTensorType::get(
+        sliceOp.getResultType().getShape(), sourceNarrowType.getElementType());
+
+    // if (!sourceWideType.getElementType().isF32() ||
```
Comment (zjgarvey) [inline compiler/src/iree/compiler/Dialect/Flow/Transforms/FusionPreprocessing.cpp]: Is this meant to be commented out?

**Review by zjgarvey — COMMENTED** [review]
Thanks for working on this, @IanWood1. 

I'm a bit worried about the generic ops produced in this pattern. For example, if the scale of the initial dequant op is very small so that the results are all very close to zero, does your generated generic op for casting f32 -> i8 just give a zero tensor?

File: compiler/src/iree/compiler/Dialect/Flow/Transforms/FusionPreprocessing.cpp (hunk 1 above)
Comment (IanWood1) [reply compiler/src/iree/compiler/Dialect/Flow/Transforms/FusionPreprocessing.cpp]: Yeah your right, this is definitely a problem. I think that solution might work, I'll take a look

**Pull request closed without merging**